integrations/mcp/src/haystack_integrations/tools/mcp/mcp_toolset.py

```diff
@@ -0,0 +1,201 @@
+            raise MCPConnectionError(message=message, server_info=self.server_info, operation="initialize") from e
+
+    def _run_sync(self, coro, timeout: float | None = None):
+        """
+        Run an async coroutine synchronously.
+
+        :param coro: The coroutine to run
+        :param timeout: Optional timeout in seconds
+        :returns: The result of the coroutine
+        :raises TimeoutError: If the operation times out
+        """
+        try:
+            # Apply timeout if specified
+            if timeout is not None:
+                coro = asyncio.wait_for(coro, timeout=timeout)
+
+            # Try to get a running loop first
+            try:
+                loop = asyncio.get_running_loop()
+            except RuntimeError:
+                # If we can't get a running loop, create a new one
+                loop = asyncio.new_event_loop()
+                asyncio.set_event_loop(loop)
+                return loop.run_until_complete(coro)
+
+            # If the loop is already running, run our coroutine inline
+            if loop.is_running():
+                # Create a new loop for our coroutine to avoid conflicts
+                # This should prevent "This event loop is already running" errors
+                temp_loop = asyncio.new_event_loop()
+                try:
+                    return temp_loop.run_until_complete(coro)
+                finally:
+                    temp_loop.close()
+            else:
+                # Use the existing loop
+                return loop.run_until_complete(coro)
+
+        except asyncio.TimeoutError as e:
+            message = f"Operation timed out after {timeout} seconds"
+            raise TimeoutError(message) from e
+        except Exception as e:
+            # Re-raise any other exceptions
+            raise e
+
+    def to_dict(self) -> dict[str, Any]:
+        """
+        Serialize the MCPToolset to a dictionary.
+
+        :returns: A dictionary representation of the MCPToolset
+        """
+        return {
+            "type": generate_qualified_class_name(type(self)),
+            "data": {
+                "server_info": self.server_info.to_dict(),
+                "connection_timeout": self.connection_timeout,
+                "invocation_timeout": self.invocation_timeout,
+            },
+        }
+
+    @classmethod
+    def from_dict(cls, data: dict[str, Any]) -> "MCPToolset":
+        """
+        Deserialize an MCPToolset from a dictionary.
+
+        :param data: Dictionary representation of the MCPToolset
+        :returns: A new MCPToolset instance
+        """
+        inner_data = data["data"]
+
+        # Reconstruct the server_info object
+        server_info_dict = inner_data.get("server_info", {})
+        server_info_class = import_class_by_name(server_info_dict["type"])
+        server_info = server_info_class.from_dict(server_info_dict)
+
+        # Create a new MCPToolset instance
+        return cls(
```

Comment:
Here we are creating a new instance of the class (which is ok), but as commented above, it will call `__init__` and so call the remote server, which can be unwanted. It will be solved if we defer the connection on a method which e.g. list tools. WDYT?

Reply:
Yeah, I wish we could. See above.